Add line folding

atgrosso opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98)
<!-- Reviewable:end -->

andrewmilson reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98)* status: 0 of 6 files reviewed, 16 unresolved discussions (waiting on @atgrosso)

___
*[`stwo_cairo_verifier/src/fri.cairo` line 1 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8EGCxnAgGIcGk0_VlS:-O8EGCxo0gh_8hJEnDB8:b-h3ip3u) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/fri.cairo#L1)):*
> ```Cairo
> pub mod folding;
> ```

Please put contents of  fri/line.cairo and fri/folding.cairo in this file.
___
*[`stwo_cairo_verifier/src/utils.cairo` line 101 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8DvofM7TCAcYeBQ-PQ:-O8DvofM7TCAcYeBQ-PR:b2emi8d) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/utils.cairo#L101)):*
> ```Cairo
> 
> pub fn bit_reverse_index(mut index: usize, mut bits: u32) -> usize {
>     assert!(bits < 32);
> ```

Isn't 32 a valid value here?

I guess it should really be 

```
use core::num::traits::BitSize;

..

assert!(bits <= usize::bits());
```
___
*[`stwo_cairo_verifier/src/utils.cairo` line 107 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8EGoGC76xwNrMhi-Y6:-O8EGoGDENI8ihdGDt2W:b-8c23dn) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/utils.cairo#L107)):*
> ```Cairo
>         result *= 2;
>         result = result | ((index / pow_of_two) & 1);
>         pow_of_two *= 2;
> ```

Saves an op



_Suggestion:_
```Cairo
        result = (result * 2) | (index & 1);
        index /= 2;
```
___
*[`stwo_cairo_verifier/src/fri/folding.cairo` line 12 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8DxJo32L18LBrpmTMA:-O8DxJo32L18LBrpmTMB:b45h) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/fri/folding.cairo#L12)):*
> ```Cairo
> 
> /// Folds a degree `d` polynomial into a degree `d/2` polynomial.
> pub fn fold_line(eval: @LineEvaluation, alpha: QM31) -> LineEvaluation {
> ```





_Suggestion:_
```Cairo
/// Folds a degree `d` polynomial into a degree `d/2` polynomial.
///
/// Let `eval` be a polynomial evaluated on line domain `E`, `alpha` be a random field
/// element and `pi(x) = 2x^2 - 1` be the circle's x-coordinate doubling map. This function
/// returns `f' = f0 + alpha * f1` evaluated on `pi(E)` such that `2f(x) = f0(pi(x)) + x *
/// f1(pi(x))`.
pub fn fold_line(eval: @LineEvaluation, alpha: QM31) -> LineEvaluation {
```
___
*[`stwo_cairo_verifier/src/fri/folding.cairo` line 14 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8EFJRyE8G9AcpfzOMM:-O8EFJRyE8G9AcpfzOMN:undefined) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/fri/folding.cairo#L14)):*
> ```Cairo
> pub fn fold_line(eval: @LineEvaluation, alpha: QM31) -> LineEvaluation {
>     let domain = eval.domain;
>     let mut values: Array<QM31> = array![];
> ```





_Suggestion:_
```Cairo
    let mut values = array![];
```
___
*[`stwo_cairo_verifier/src/fri/folding.cairo` line 26 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8EFMWwBFP1zXN4iWlv:-O8EFMWwBFP1zXN4iWlw:undefined) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/fri/folding.cairo#L26)):*
> ```Cairo
>     LineEvaluationImpl::new(domain.double(), values)
> }
> pub fn ibutterfly(v0: QM31, v1: QM31, itwid: M31) -> (QM31, QM31) {
> ```





_Suggestion:_
```Cairo
}

pub fn ibutterfly(v0: QM31, v1: QM31, itwid: M31) -> (QM31, QM31) {
```
___
*[`stwo_cairo_verifier/src/fri/folding.cairo` line 27 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8E8WfP9bAnfC-nWQBh:-O8E8WfP9bAnfC-nWQBi:undefined) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/fri/folding.cairo#L27)):*
> ```Cairo
> }
> pub fn ibutterfly(v0: QM31, v1: QM31, itwid: M31) -> (QM31, QM31) {
>     (v0 + v1, (v0 - v1) * itwid.into())
> ```





_Suggestion:_
```Cairo
).mul_m31(itwid)
```
___
*[`stwo_cairo_verifier/src/fri/folding.cairo` line 29 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8Du9B4ECbkYcLyE1ck:-O8Du9B4ECbkYcLyE1cl:undefined) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/fri/folding.cairo#L29)):*
> ```Cairo
>     (v0 + v1, (v0 - v1) * itwid.into())
> }
> mod test {
> ```





_Suggestion:_
```Cairo
}

#[cfg(test)]
mod test {
```
___
*[`stwo_cairo_verifier/src/fri/folding.cairo` line 40 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8Dt3DA8_TTvU2k_9o2:-O8Dt3DA8_TTvU2k_9o3:undefined) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/fri/folding.cairo#L40)):*
> ```Cairo
>     use stwo_cairo_verifier::utils::{bit_reverse_index, pow};
>     use stwo_cairo_verifier::fri::folding::{FOLD_STEP, CIRCLE_TO_LINE_FOLD_STEP};
>     #[test]
> ```





_Suggestion:_
```Cairo
    use stwo_cairo_verifier::fri::folding::{FOLD_STEP, CIRCLE_TO_LINE_FOLD_STEP};

    #[test]
```
___
*[`stwo_cairo_verifier/src/fri/folding.cairo` line 53 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8EAwr3EahL4qiZ9VxI:-O8EAwr3EahL4qiZ9VxJ:undefined) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/fri/folding.cairo#L53)):*
> ```Cairo
>         let result = sparse_line_evaluation.fold(alpha);
>         let expected_result = array![qm31(515899232, 1030391528, 1006544539, 11142505)];
>         assert_eq!(expected_result, result);
> ```





_Suggestion:_
```Cairo
        let alpha = qm31(1047716961, 506143067, 1065078409, 990356067);

        let result = sparse_line_evaluation.fold(alpha);

        assert_eq!(result, array![qm31(515899232, 1030391528, 1006544539, 11142505)]);
```
___
*[`stwo_cairo_verifier/src/fri/folding.cairo` line 55 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8DtE-nCFS5ztz5NXjT:-O8DtE-nCFS5ztz5NXjU:undefined) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/fri/folding.cairo#L55)):*
> ```Cairo
>         assert_eq!(expected_result, result);
>     }
>     #[test]
> ```





_Suggestion:_
```Cairo
    }

    #[test]
```
___
*[`stwo_cairo_verifier/src/fri/folding.cairo` line 68 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8DxyNW8LBVlX_a28An:-O8DxyNW8LBVlX_a28Ao:b45h) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/fri/folding.cairo#L68)):*
> ```Cairo
>         let result = sparse_line_evaluation.fold(alpha);
>         let expected_result = array![qm31(1379727866, 1083096056, 1409020369, 1977903500)];
>         assert_eq!(expected_result, result);
> ```





_Suggestion:_
```Cairo
        let alpha = qm31(2084521793, 1326105747, 548635876, 1532708504);
    
        let result = sparse_line_evaluation.fold(alpha);
    
        assert_eq!(result, array![qm31(1379727866, 1083096056, 1409020369, 1977903500)]);
```
___
*[`stwo_cairo_verifier/src/poly/line.cairo` line 17 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8ED6Oh-UTrxihroAXx:-O8ED6Oh-UTrxihroAXy:b5zk7f3) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/poly/line.cairo#L17)):*
> ```Cairo
>     /// Coefficients of the polynomial in [line_ifft] algorithm's basis.
>     ///
>     /// The coefficients are stored in bit-reversed order.
> ```

Given line_ifft not here



_Suggestion:_
```Cairo
/// The coefficients of the polynomial stored in bit-reversed order.
/// 
/// The coefficients are in a basis relating to the circle's x-coordinate doubling 
/// map `pi(x) = 2x^2 - 1` i.e.
///
/// ```text
/// B = { 1 } * { x } * { pi(x) } * { pi(pi(x)) } * ...
///   = { 1, x, pi(x), pi(x) * x, pi(pi(x)), pi(pi(x)) * x, pi(pi(x)) * pi(x), ... }
/// ```
```
___
*[`stwo_cairo_verifier/src/poly/line.cairo` line 57 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8EDVt--I6ksH72vked:-O8EC5cg4iAs5xbj_aqP:b-zhceqv) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/poly/line.cairo#L57)):*
> ```Cairo
>     /// Returns a domain comprising of the x-coordinates of points in a coset.
>     fn new(coset: Coset) -> LineDomain {
>         LineDomain { coset: coset }
> ```

Please include the asserts.
___
*[`stwo_cairo_verifier/src/poly/line.cairo` line 98 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8EDlT_7btcncMdn7f9:-O8EDlT_7btcncMdn7fA:b-d15vhc) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/poly/line.cairo#L98)):*
> ```Cairo
> 
> /// Holds a small foldable subset of univariate SecureField polynomial evaluations.
> /// Evaluation is held at the CPU backend.
> ```

Verifier doesn't require a concept of backend
___
*[`stwo_cairo_verifier/src/poly/line.cairo` line 115 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8EEWXC8-CwA9nckh3X:-O8EEWXC8-CwA9nckh3Y:b45h) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/poly/line.cairo#L115)):*
> ```Cairo
>         };
>         return res;
>     }
> ```





_Suggestion:_
```Cairo
    fn fold(self: SparseLineEvaluation, alpha: QM31) -> Array<QM31> {
        let mut res = array![];
        for eval in self.subline_evals {
            res.append(fold_line(eval, alpha)[0])
        }
        res
    }
```


<!-- Sent from Reviewable.io -->

andrewmilson reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98)* status: 0 of 6 files reviewed, 16 unresolved discussions (waiting on @atgrosso)

___
*[`stwo_cairo_verifier/src/utils.cairo` line 107 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8EGoGC76xwNrMhi-Y6:-O8EHQXs2HD1yMPGI96z:b-cw0a6y) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/utils.cairo#L107)):*
<details><summary><i>Previously, andrewmilson (Andrew Milson) wrote…</i></summary><blockquote>

Saves an op
</blockquote></details>

Actually this might be better

```
use core::traits::DivRem;
..
const NZ2: NonZero<usize> = 2;
..
let (next_index, bit) = DivRem::div_rem(index, NZ2);
result = (result * 2) | bit;
index = next_index;
```


<!-- Sent from Reviewable.io -->

atgrosso commented:
> > _[Reviewable](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98)_ status: 0 of 6 files reviewed, 16 unresolved discussions (waiting on @atgrosso)
> 
> _[`stwo_cairo_verifier/src/fri.cairo` line 1 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8EGCxnAgGIcGk0_VlS:-O8EGCxo0gh_8hJEnDB8:b-h3ip3u) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/fri.cairo#L1)):_
> 
> > ```cairo
> > pub mod folding;
> > ```
> 
> Please put contents of fri/line.cairo and fri/folding.cairo in this file.
> 
> _[`stwo_cairo_verifier/src/utils.cairo` line 101 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8DvofM7TCAcYeBQ-PQ:-O8DvofM7TCAcYeBQ-PR:b2emi8d) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/utils.cairo#L101)):_
> 
> > ```cairo
> > pub fn bit_reverse_index(mut index: usize, mut bits: u32) -> usize {
> >     assert!(bits < 32);
> > ```
> 
> Isn't 32 a valid value here?
> 
> I guess it should really be
> 
> ```
> use core::num::traits::BitSize;
> 
> ..
> 
> assert!(bits <= usize::bits());
> ```
You are right, using BitSize::bits() bound exactly to usize. Done!
> 
> _[`stwo_cairo_verifier/src/utils.cairo` line 107 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8EGoGC76xwNrMhi-Y6:-O8EGoGDENI8ihdGDt2W:b-8c23dn) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/utils.cairo#L107)):_
> 
> > ```cairo
> >         result *= 2;
> >         result = result | ((index / pow_of_two) & 1);
> >         pow_of_two *= 2;
> > ```
> 
> Saves an op
> 
> _Suggestion:_
> 
> ```cairo
>         result = (result * 2) | (index & 1);
>         index /= 2;
> ```
> 
Good catch. Done!
> _[`stwo_cairo_verifier/src/fri/folding.cairo` line 12 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8DxJo32L18LBrpmTMA:-O8DxJo32L18LBrpmTMB:b45h) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/fri/folding.cairo#L12)):_
> 
> > ```cairo
> > /// Folds a degree `d` polynomial into a degree `d/2` polynomial.
> > pub fn fold_line(eval: @LineEvaluation, alpha: QM31) -> LineEvaluation {
> > ```
> 
> _Suggestion:_
> 
> ```cairo
> /// Folds a degree `d` polynomial into a degree `d/2` polynomial.
> ///
> /// Let `eval` be a polynomial evaluated on line domain `E`, `alpha` be a random field
> /// element and `pi(x) = 2x^2 - 1` be the circle's x-coordinate doubling map. This function
> /// returns `f' = f0 + alpha * f1` evaluated on `pi(E)` such that `2f(x) = f0(pi(x)) + x *
> /// f1(pi(x))`.
> pub fn fold_line(eval: @LineEvaluation, alpha: QM31) -> LineEvaluation {
> ```
> 
Done
> _[`stwo_cairo_verifier/src/fri/folding.cairo` line 14 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8EFJRyE8G9AcpfzOMM:-O8EFJRyE8G9AcpfzOMN:undefined) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/fri/folding.cairo#L14)):_
> 
> > ```cairo
> > pub fn fold_line(eval: @LineEvaluation, alpha: QM31) -> LineEvaluation {
> >     let domain = eval.domain;
> >     let mut values: Array<QM31> = array![];
> > ```
> 
> _Suggestion:_
> 
> ```cairo
>     let mut values = array![];
> ```
> 
Done
> _[`stwo_cairo_verifier/src/fri/folding.cairo` line 26 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8EFMWwBFP1zXN4iWlv:-O8EFMWwBFP1zXN4iWlw:undefined) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/fri/folding.cairo#L26)):_
> 
> > ```cairo
> >     LineEvaluationImpl::new(domain.double(), values)
> > }
> > pub fn ibutterfly(v0: QM31, v1: QM31, itwid: M31) -> (QM31, QM31) {
> > ```
> 
> _Suggestion:_
> 
> ```cairo
> }
> 
> pub fn ibutterfly(v0: QM31, v1: QM31, itwid: M31) -> (QM31, QM31) {
> ```
> 
Done
> _[`stwo_cairo_verifier/src/fri/folding.cairo` line 27 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8E8WfP9bAnfC-nWQBh:-O8E8WfP9bAnfC-nWQBi:undefined) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/fri/folding.cairo#L27)):_
> 
> > ```cairo
> > }
> > pub fn ibutterfly(v0: QM31, v1: QM31, itwid: M31) -> (QM31, QM31) {
> >     (v0 + v1, (v0 - v1) * itwid.into())
> > ```
> 
> _Suggestion:_
> 
> ```cairo
> ).mul_m31(itwid)
> ```
> 
Added
> _[`stwo_cairo_verifier/src/fri/folding.cairo` line 29 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8Du9B4ECbkYcLyE1ck:-O8Du9B4ECbkYcLyE1cl:undefined) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/fri/folding.cairo#L29)):_
> 
> > ```cairo
> >     (v0 + v1, (v0 - v1) * itwid.into())
> > }
> > mod test {
> > ```
> 
> _Suggestion:_
> 
> ```cairo
> }
> 
> #[cfg(test)]
> mod test {
> ```
> 
Done
> _[`stwo_cairo_verifier/src/fri/folding.cairo` line 40 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8Dt3DA8_TTvU2k_9o2:-O8Dt3DA8_TTvU2k_9o3:undefined) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/fri/folding.cairo#L40)):_
> 
> > ```cairo
> >     use stwo_cairo_verifier::utils::{bit_reverse_index, pow};
> >     use stwo_cairo_verifier::fri::folding::{FOLD_STEP, CIRCLE_TO_LINE_FOLD_STEP};
> >     #[test]
> > ```
> 
> _Suggestion:_
> 
> ```cairo
>     use stwo_cairo_verifier::fri::folding::{FOLD_STEP, CIRCLE_TO_LINE_FOLD_STEP};
> 
>     #[test]
> ```
> 
Done
> _[`stwo_cairo_verifier/src/fri/folding.cairo` line 53 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8EAwr3EahL4qiZ9VxI:-O8EAwr3EahL4qiZ9VxJ:undefined) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/fri/folding.cairo#L53)):_
> 
> > ```cairo
> >         let result = sparse_line_evaluation.fold(alpha);
> >         let expected_result = array![qm31(515899232, 1030391528, 1006544539, 11142505)];
> >         assert_eq!(expected_result, result);
> > ```
> 
> _Suggestion:_
> 
> ```cairo
>         let alpha = qm31(1047716961, 506143067, 1065078409, 990356067);
> 
>         let result = sparse_line_evaluation.fold(alpha);
> 
>         assert_eq!(result, array![qm31(515899232, 1030391528, 1006544539, 11142505)]);
> ```
> 
Done
> _[`stwo_cairo_verifier/src/fri/folding.cairo` line 55 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8DtE-nCFS5ztz5NXjT:-O8DtE-nCFS5ztz5NXjU:undefined) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/fri/folding.cairo#L55)):_
> 
> > ```cairo
> >         assert_eq!(expected_result, result);
> >     }
> >     #[test]
> > ```
> 
> _Suggestion:_
> 
> ```cairo
>     }
> 
>     #[test]
> ```
>
Done 
> _[`stwo_cairo_verifier/src/fri/folding.cairo` line 68 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8DxyNW8LBVlX_a28An:-O8DxyNW8LBVlX_a28Ao:b45h) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/fri/folding.cairo#L68)):_
> 
> > ```cairo
> >         let result = sparse_line_evaluation.fold(alpha);
> >         let expected_result = array![qm31(1379727866, 1083096056, 1409020369, 1977903500)];
> >         assert_eq!(expected_result, result);
> > ```
> 
> _Suggestion:_
> 
> ```cairo
>         let alpha = qm31(2084521793, 1326105747, 548635876, 1532708504);
>     
>         let result = sparse_line_evaluation.fold(alpha);
>     
>         assert_eq!(result, array![qm31(1379727866, 1083096056, 1409020369, 1977903500)]);
> ```
> 
Done
> _[`stwo_cairo_verifier/src/poly/line.cairo` line 17 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8ED6Oh-UTrxihroAXx:-O8ED6Oh-UTrxihroAXy:b5zk7f3) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/poly/line.cairo#L17)):_
> 
> > ```cairo
> >     /// Coefficients of the polynomial in [line_ifft] algorithm's basis.
> >     ///
> >     /// The coefficients are stored in bit-reversed order.
> > ```
> 
> Given line_ifft not here
> 
> _Suggestion:_
> 
> ```cairo
> /// The coefficients of the polynomial stored in bit-reversed order.
> /// 
> /// The coefficients are in a basis relating to the circle's x-coordinate doubling 
> /// map `pi(x) = 2x^2 - 1` i.e.
> ///
> /// ```text
> /// B = { 1 } * { x } * { pi(x) } * { pi(pi(x)) } * ...
> ///   = { 1, x, pi(x), pi(x) * x, pi(pi(x)), pi(pi(x)) * x, pi(pi(x)) * pi(x), ... }
> /// ```
> ```
> 
Yes! good catch. sgtm
> _[`stwo_cairo_verifier/src/poly/line.cairo` line 57 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8EDVt--I6ksH72vked:-O8EC5cg4iAs5xbj_aqP:b-zhceqv) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/poly/line.cairo#L57)):_
> 
> > ```cairo
> >     /// Returns a domain comprising of the x-coordinates of points in a coset.
> >     fn new(coset: Coset) -> LineDomain {
> >         LineDomain { coset: coset }
> > ```
> 
> Please include the asserts.
> 
Done. The tests that use odds are refactored in the next pr (odds is added there)
> _[`stwo_cairo_verifier/src/poly/line.cairo` line 98 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8EDlT_7btcncMdn7f9:-O8EDlT_7btcncMdn7fA:b-d15vhc) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/poly/line.cairo#L98)):_
> 
> > ```cairo
> > /// Holds a small foldable subset of univariate SecureField polynomial evaluations.
> > /// Evaluation is held at the CPU backend.
> > ```
> 
> Verifier doesn't require a concept of backend
> 
Oops, removed it
> _[`stwo_cairo_verifier/src/poly/line.cairo` line 115 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8EEWXC8-CwA9nckh3X:-O8EEWXC8-CwA9nckh3Y:b45h) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/poly/line.cairo#L115)):_
> 
> > ```cairo
> >         };
> >         return res;
> >     }
> > ```
> 
> _Suggestion:_
> 
> ```cairo
>     fn fold(self: SparseLineEvaluation, alpha: QM31) -> Array<QM31> {
>         let mut res = array![];
>         for eval in self.subline_evals {
>             res.append(fold_line(eval, alpha)[0])
>         }
>         res
>     }
> ```
Done

atgrosso commented:
> > _[Reviewable](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98)_ status: 0 of 6 files reviewed, 16 unresolved discussions (waiting on @atgrosso)
> 
> _[`stwo_cairo_verifier/src/utils.cairo` line 107 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8EGoGC76xwNrMhi-Y6:-O8EHQXs2HD1yMPGI96z:b-cw0a6y) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/utils.cairo#L107)):_
> Previously, andrewmilson (Andrew Milson) wrote…
> 
> Actually this might be better
> 
> ```
> use core::traits::DivRem;
> ..
> const NZ2: NonZero<usize> = 2;
> ..
> let (next_index, bit) = DivRem::div_rem(index, NZ2);
> result = (result * 2) | bit;
> index = next_index;
> ```

It looks better to me too. Done

atgrosso commented:
> _[`stwo_cairo_verifier/src/fri.cairo` line 1 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8EGCxnAgGIcGk0_VlS:-O8EGCxo0gh_8hJEnDB8:b-h3ip3u) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/fri.cairo#L1)):_
> 
> > ```cairo
> > pub mod folding;
> > ```
> 
> Please put contents of fri/line.cairo and fri/folding.cairo in this file.
> 
Done. Currently we only moved folding.cairo on this PR. We'll continue to move the missing files and functions related to fri

andrewmilson reviewed (approved):
Thanks. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with couple small comments

> Reviewed 2 of 6 files at r1, 3 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-:-O8SzsA68VE3ZPdyDuk4:bz7t4br)* status: 5 of 8 files reviewed, 18 unresolved discussions (waiting on @atgrosso)

___
*[`stwo_cairo_verifier/src/circle.cairo` line 44 at r2](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8Sw6bvBPquExIZEonB:-O8Sw6bvBPquExIZEonC:b-59rn4h) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/7878534afe1d4ae430b42e528a7891d37cd7c9ac/stwo_cairo_verifier/src/circle.cairo#L44)):*
> ```Cairo
>         let sx = x.clone() * x.clone();
>         sx.clone() + sx - M31One::one()
>     }
> ```

The example acts as a Rust docstring test so best to remove.
Also clone can be omitted.



_Suggestion:_
```Cairo
    /// Applies the circle's x-coordinate doubling map.
    fn double_x(x: M31) -> M31 {
        let sqx = x * x;
        sqx + sqx - M31One::one()
    }
```
___
*[`stwo_cairo_verifier/src/circle.cairo` line 67 at r2](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8SwTscDbVof6HW6Xwe:-O8SwTscDbVof6HW6Xwf:brhumzt) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/7878534afe1d4ae430b42e528a7891d37cd7c9ac/stwo_cairo_verifier/src/circle.cairo#L67)):*
> ```Cairo
>         };
>         res
>     }
> ```

Same here.



_Suggestion:_
```Cairo
    /// Returns the log order of a point.
    ///
    /// All points have an order of the form `2^k`.
    fn log_order(self: @CirclePointM31) -> u32 {
        // we only need the x-coordinate to check order since the only point
        // with x=1 is the circle's identity
        let mut res = 0;
        let mut cur = self.x.clone();
        while cur != M31One::one() {
            cur = Self::double_x(cur);
            res += 1;
        };
        res
    }
```
___
*[`stwo_cairo_verifier/src/poly/line.cairo` line 115 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8EEWXC8-CwA9nckh3X:-O8SyoWC3hwPLqkslDfX:b-up6ffe) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/poly/line.cairo#L115)):*
<details><summary><i>Previously, atgrosso (Tomás Grosso) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Don't see any change


<!-- Sent from Reviewable.io -->

atgrosso commented:
> Thanks. <img alt=":lgtm:" width="61" height="20" src="https://camo.githubusercontent.com/204e5a43dbe1a0ce96b00dc9f6efaa7c31943c30f6d7fc9f92fa23ad865d6ea0/68747470733a2f2f72657669657761626c652e696f2f6c67746d2e706e67"> with couple small comments
> 
> > Reviewed 2 of 6 files at r1, 3 of 6 files at r2, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-:-O8SzsA68VE3ZPdyDuk4:bz7t4br)_ status: 5 of 8 files reviewed, 18 unresolved discussions (waiting on @atgrosso)
> 
> _[`stwo_cairo_verifier/src/circle.cairo` line 44 at r2](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8Sw6bvBPquExIZEonB:-O8Sw6bvBPquExIZEonC:b-59rn4h) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/7878534afe1d4ae430b42e528a7891d37cd7c9ac/stwo_cairo_verifier/src/circle.cairo#L44)):_
> 
> > ```cairo
> >         let sx = x.clone() * x.clone();
> >         sx.clone() + sx - M31One::one()
> >     }
> > ```
> 
> The example acts as a Rust docstring test so best to remove. Also clone can be omitted.
> 
> _Suggestion:_
> 
> ```cairo
>     /// Applies the circle's x-coordinate doubling map.
>     fn double_x(x: M31) -> M31 {
>         let sqx = x * x;
>         sqx + sqx - M31One::one()
>     }
> ```
> 
Done
> _[`stwo_cairo_verifier/src/circle.cairo` line 67 at r2](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8SwTscDbVof6HW6Xwe:-O8SwTscDbVof6HW6Xwf:brhumzt) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/7878534afe1d4ae430b42e528a7891d37cd7c9ac/stwo_cairo_verifier/src/circle.cairo#L67)):_
> 
> > ```cairo
> >         };
> >         res
> >     }
> > ```
> 
> Same here.
> 
> _Suggestion:_
> 
> ```cairo
>     /// Returns the log order of a point.
>     ///
>     /// All points have an order of the form `2^k`.
>     fn log_order(self: @CirclePointM31) -> u32 {
>         // we only need the x-coordinate to check order since the only point
>         // with x=1 is the circle's identity
>         let mut res = 0;
>         let mut cur = self.x.clone();
>         while cur != M31One::one() {
>             cur = Self::double_x(cur);
>             res += 1;
>         };
>         res
>     }
> ```
> 
Done
> _[`stwo_cairo_verifier/src/poly/line.cairo` line 115 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/98#-O8EEWXC8-CwA9nckh3X:-O8SyoWC3hwPLqkslDfX:b-up6ffe) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/cc58115e8f412a48893e6760227bc73e66afa0cf/stwo_cairo_verifier/src/poly/line.cairo#L115)):_
> Previously, atgrosso (Tomás Grosso) wrote…
> 
> > Done
> 
> Don't see any change

Sry, i only removed the "return". I didn't see the `for` cycle.
I tried to use a `for` but scarb failed with: "not yet implemented: Handle usages of `for` loop expression"
Maybe the scarb version (nightly-2024-06-15) used don't support the for structure?

Thankss
